### PR TITLE
[KGE] Add evaluate logic

### PR DESCRIPTION
**Pull request by nshah-sc**
Add code which computes the evaluation metrics for KGE training.

Main entrypoint is `evaluate` which just calls other private functions for organization.

Where is the documentation for this feature?: N/A

Did you add automated tests or write a test plan?

***Updated Changelog.md?*** NO

***Ready for code review?:*** YES

### CODE REVIEW

File: python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    step_count = 0
+
+    # Process validation batches
+    while True:
+        try:
+            batch_loss, logits, labels, edge_types = pipeline.progress(val_iter)
+            losses.append(batch_loss.item())
+
+            # Accumulate metrics for each edge type in this batch
+            for edge_type in unique_edge_types:
```
Comment (svij-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py:180]: is there a chance that a specific edge type doesnt occur in a batch?
Or does on one rank in a batch but not in other?

If so, does that affect metric computations?

Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py]: good point -- was not considered rigorously in the context of this experimental code, but updated for reasonable behavior in this setting (falling back to the results from other ranks)

File: python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            break
+
+    logger.info(f"Completed {phase} evaluation over {step_count} steps.")
+
+    # Aggregate metrics and prepare for distributed reduction
+    aggregated_metrics = _aggregate_metrics(accumulators, unique_edge_types, device)
+
+    # Perform distributed reduction to average across all workers
+    for metric in aggregated_metrics:
+        dist.all_reduce(metric, op=dist.ReduceOp.AVG)
```
[the remark this answers is not in the document]
Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py]: agreed; this is now the protocol

File: python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        evaluation_phase_config: Configuration specifying evaluation parameters
+            like hit_rates_at_k values
+        graph_metadata: Metadata containing information about condensed edge types
+            in the graph
+
+    Returns:
+        A tuple containing:
+        - overall_loss: Average loss across all batches and edge types
+        - metrics_by_edge_type: Dictionary mapping each CondensedEdgeType to
+          a tuple of (avg_pos_score, avg_neg_score, avg_mrr, avg_hit_rates)
```
[the remark this answers is not in the document]
Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py]: good call; cleaned this up

File: python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+    # Perform distributed reduction on all metric tensors
+    accumulator.sum_metrics_over_ranks()
+
+    # Compute final averaged metrics and format results
+    result = accumulator.compute_final_metrics(unique_edge_types=unique_edge_types)
+
+    # Restore original model state
+    pipeline._model.module.set_phase(original_phase)
+    pipeline._model.train()
```
Comment (svij-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/evaluation.py]: only enable training if "was_training" before?
